platform-sdk/docs/proposals/healthMonitor/healthMonitorProposal.md

```diff
@@ -0,0 +1,127 @@
+# Health Monitor Proposal
+
+---
+
+## Summary
+
+A replacement for the current hard backpressure mechanism using a health monitor.
+
+| Metadata           | Entities                            | 
+|--------------------|-------------------------------------|
+| Designers          | Cody Littley                        |
+| Functional Impacts | Consensus node performance          |
+| Related Proposals  | n/a                                 |
+| HIPS               | n/a                                 |
+
+---
+
+## Purpose and Context
+
+The platform's current backpressure mechanism has very bad performance characteristics.
+We need to replace the current backpressure implementation with something more performant.
+
+### Requirements
+
+- The backpressure mechanism must prevent an OOM in the event of any arbitrary component
+  becoming arbitrarily slow.
+- The backpressure mechanism should not slow down critical processes when enganged.
+  Specifically, transaction handling should not be impeded when backpressure is engaged.
+  This includes slowdowns due to the transaction handling thread being starved of work.
+- The backpressure mechanism should not be overly aggressive. That is, we shouldn't see
+  it engaging under normal circumstances on nodes that are not in a state of distress.
+- The implementation of the backpressure mechanism should be sufficiently clean and
+  maintainable. A performant solution that is too complex is not a solution we should
+  accept as a solution.
+
+### Design Decisions
+
+This change can be split into two distinct sub-problems: the detection of a system in duress, and the action that
+is taken to apply backpressure on a distressed system.
+
+#### Detection
+
+The current platform is built from a number of "components". A component is usually a unit of logic that runs on its
+own thread (or logical thread). Each component has a queue (or logical queue) of data it is ingesting. If a component
+produces output, that output is enqueued within the queues of other components.
+
+For each component with a queue, define a preferred size. The queue will not apply hard backpressure even if the
+size of the queue exceeds the preferred size. That is, insertion into the queue will never block. However, a queue
+whose size exceeds its preferred size will be considered to be "unhealthy".
+
+In the background, a new component called the "health monitor" will periodically check the health of each queue.
+Checking the health of a queue is cheap (i.e. an atomic long read). The frequency of polling is configurable, with the
+default being 10hz.
+
+The health monitor records the length of time that each queue has been in an unhealthy state. If a queue has been
+unhealthy for one minute, that is a much stronger signal than if a queue has been unhealthy for a few hundred
+milliseconds. If a previously unhealthy queue is observed as being healthy, the timer for that queue is immediately
+reset to 0.
```

Comment:
What is being used as a timer and what frequency does that "clock" operate at?

Reply:
I want to decouple the monitoring thread from that. Sounds like a bad idea to have monitoring and execution using the same thread pool (if I understand it correctly)